File: web_api_tests/csharp/LoadTest.cs
using System;
using System.Net.Http;
using System.Collections.Generic;
using System.Threading.Tasks;
using System.Diagnostics;

class LoadTest
{
    private static readonly string BASE_URL = "http://localhost:3002";
    private static readonly HttpClient client = new();

    static async Task Main()
    {
        Console.WriteLine("\n========================================");
        Console.WriteLine("  C#/.NET - LOAD TEST");
        Console.WriteLine("========================================\n");

        var startTime = Stopwatch.GetTimestamp();

        // Create users
        Console.WriteLine("Creating 100 users...");
        var userIds = new List<int>();
        for (int i = 0; i < 100; i++)
        {
            var res = await MakeRequest("POST", "/api/users", new
            {
                username = $"user{i}",
                email = $"user{i}@example.com",
                displayName = $"User {i}",
            });
            if (res != null && res.ContainsKey("id"))
                userIds.Add((int)(long)res["id"]);
        }

        // Create posts
        Console.WriteLine("Creating 500 posts...");
        var postIds = new List<int>();
        for (int i = 0; i < 500; i++)
        {
            var res = await MakeRequest("POST", "/api/posts", new
            {
                userId = userIds[i % userIds.Count],
                content = $"Post #{i} - Lorem ipsum dolor sit amet",
            });
            if (res != null && res.ContainsKey("id"))
                postIds.Add((int)(long)res["id"]);
        }

        // Add comments
        Console.WriteLine("Adding 1000 comments...");
        for (int i = 0; i < 1000; i++)
        {
            await MakeRequest("POST", "/api/comments", new
            {
                postId = postIds[i % postIds.Count],
                userId = userIds[i % userIds.Count],
                text = $"Comment #{i} - Great post!",
            });
        }

        // Like posts
        Console.WriteLine("Liking 2000 times...");
        int likeCount = 0;
        for (int i = 0; i < 2000; i++)
        {
            var res = await MakeRequest("POST", "/api/likes", new
            {
                postId = postIds[i % postIds.Count],
                userId = userIds[(i + 1) % userIds.Count],
            });
            if (res != null && res.ContainsKey("success"))
                likeCount++;
        }

        // Follow users
        Console.WriteLine("Creating 500 follow relationships...");
        int followCount = 0;
        for (int i = 0; i < 500; i++)
        {
            var res = await MakeRequest("POST", "/api/follow", new
            {
                followerId = userIds[i % userIds.Count],
                followingId = userIds[(i + 1) % userIds.Count],
            });
            if (res != null && res.ContainsKey("success"))
                followCount++;
        }

        // Fetch feeds
        Console.WriteLine("Fetching 100 feeds...");
        for (int i = 0; i < 100; i++)
        {
            await client.GetAsync($"{BASE_URL}/api/users/{userIds[i]}/feed");
        }

        // Fetch posts with comments
        Console.WriteLine("Fetching posts with comments...");
        for (int i = 0; i < 100; i++)
        {
            await client.GetAsync($"{BASE_URL}/api/posts/{postIds[i]}");
            await client.GetAsync($"{BASE_URL}/api/posts/{postIds[i]}/comments");
        }

        var totalTime = Stopwatch.GetElapsedTime(startTime).TotalMilliseconds;

        Console.WriteLine("\n========================================");
        Console.WriteLine("           LOAD TEST RESULTS");
        Console.WriteLine("========================================");
        Console.WriteLine($"Users Created:        100");
        Console.WriteLine($"Posts Created:        500");
        Console.WriteLine($"Comments Added:       1000");
        Console.WriteLine($"Likes Successful:     {likeCount}");
        Console.WriteLine($"Follows Successful:   {followCount}");
        Console.WriteLine($"Total Execution Time: {totalTime:F0} ms");
        Console.WriteLine($"Requests/sec:         {(5000 * 1000 / totalTime):F2}");
        Console.WriteLine("========================================\n");
    }

    static async Task<Dictionary<string, object>> MakeRequest(string method, string path, object body)
    {
        try
        {
            var request = new HttpRequestMessage(new HttpMethod(method), $"{BASE_URL}{path}");
            if (body != null)
            {
                var json = System.Text.Json.JsonSerializer.Serialize(body);
                request.Content = new StringContent(json, System.Text.Encoding.UTF8, "application/json");
            }
            var response = await client.SendAsync(request);
            var content = await response.Content.ReadAsStringAsync();
            return System.Text.Json.JsonSerializer.Deserialize<Dictionary<string, object>>(content);
        }
        catch { return null; }
    }
}

File: web_api_tests/python/load_test.py
import asyncio
import aiohttp
import time

BASE_URL = "http://localhost:3001"

async def make_request(session, method, path, json=None):
    async with session.request(method, f"{BASE_URL}{path}", json=json) as resp:
        try:
            data = await resp.json()
        except:
            data = None
        return {"status": resp.status, "data": data}

async def run_load_test():
    print("\n========================================")
    print("  PYTHON/FASTAPI - LOAD TEST")
    print("========================================\n")

    start_time = time.time()

    async with aiohttp.ClientSession() as session:
        # Create users
        print("Creating 100 users...")
        user_ids = []
        for i in range(100):
            res = await make_request(session, "POST", "/api/users", {
                "username": f"user{i}",
                "email": f"user{i}@example.com",
                "displayName": f"User {i}",
            })
            user_ids.append(res["data"]["id"])

        # Create posts
        print("Creating 500 posts...")
        post_ids = []
        for i in range(500):
            user_id = user_ids[i % len(user_ids)]
            res = await make_request(session, "POST", "/api/posts", {
                "userId": user_id,
                "content": f"Post #{i} - Lorem ipsum dolor sit amet",
            })
            post_ids.append(res["data"]["id"])

        # Add comments
        print("Adding 1000 comments...")
        for i in range(1000):
            post_id = post_ids[i % len(post_ids)]
            user_id = user_ids[i % len(user_ids)]
            await make_request(session, "POST", "/api/comments", {
                "postId": post_id,
                "userId": user_id,
                "text": f"Comment #{i} - Great post!",
            })

        # Like posts
        print("Liking 2000 times...")
        like_count = 0
        for i in range(2000):
            post_id = post_ids[i % len(post_ids)]
            user_id = user_ids[(i + 1) % len(user_ids)]
            res = await make_request(session, "POST", "/api/likes", {
                "postId": post_id,
                "userId": user_id,
            })
            if res["status"] == 201:
                like_count += 1

        # Follow users
        print("Creating 500 follow relationships...")
        follow_count = 0
        for i in range(500):
            follower_id = user_ids[i % len(user_ids)]
            following_id = user_ids[(i + 1) % len(user_ids)]
            res = await make_request(session, "POST", "/api/follow", {
                "followerId": follower_id,
                "followingId": following_id,
            })
            if res["status"] == 201:
                follow_count += 1

        # Fetch feeds
        print("Fetching 100 feeds...")
        for i in range(100):
            await make_request(session, "GET", f"/api/users/{user_ids[i]}/feed")

        # Fetch posts with comments
        print("Fetching posts with comments...")
        for i in range(100):
            await make_request(session, "GET", f"/api/posts/{post_ids[i]}")
            await make_request(session, "GET", f"/api/posts/{post_ids[i]}/comments")

    total_time = (time.time() - start_time) * 1000

    print("\n========================================")
    print("           LOAD TEST RESULTS")
    print("========================================")
    print(f"Users Created:        100")
    print(f"Posts Created:        500")
    print(f"Comments Added:       1000")
    print(f"Likes Successful:     {like_count}")
    print(f"Follows Successful:   {follow_count}")
    print(f"Total Execution Time: {total_time:.0f} ms")
    print(f"Requests/sec:         {(5000 * 1000 / total_time):.2f}")
    print("========================================\n")

if __name__ == "__main__":
    asyncio.run(run_load_test())
